components/motor/gpio/loop.go

```diff
@@ -60,27 +75,259 @@ func (m *EncodedMotor) setupControlLoop() error {
+
+	controlLoopConfig control.Config
+	blockNames        map[string][]string
+	loop              *control.Loop
+}
+
+// SetPower sets the percentage of power the motor should employ between -1 and 1.
+// Negative power implies a backward directional rotational.
+func (cm *controlledMotor) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	cm.opMgr.CancelRunning(ctx)
+	if cm.loop != nil {
+		cm.loop.Pause()
+	}
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	return cm.real.SetPower(ctx, powerPct, nil)
+}
+
+// IsPowered returns whether or not the motor is currently on, and the percent power (between 0
+// and 1, if the motor is off then the percent power will be 0).
+func (cm *controlledMotor) IsPowered(ctx context.Context, extra map[string]interface{}) (bool, float64, error) {
+	return cm.real.IsPowered(ctx, extra)
+}
+
+// IsMoving returns if the motor is moving or not.
+func (cm *controlledMotor) IsMoving(ctx context.Context) (bool, error) {
+	return cm.real.IsMoving(ctx)
+}
+
+// Stop stops rpmMonitor and stops the real motor.
+func (cm *controlledMotor) Stop(ctx context.Context, extra map[string]interface{}) error {
+	// after the motor is created, Stop is called, but if the PID controller
+	// is auto-tuning, the loop needs to keep running
+	if cm.loop != nil && !cm.loop.GetTuning(ctx) {
+		cm.loop.Pause()
+	}
+	return cm.real.Stop(ctx, nil)
+}
+
+// Close cleanly shuts down the motor.
+func (cm *controlledMotor) Close(ctx context.Context) error {
+	if err := cm.Stop(ctx, nil); err != nil {
+		return err
+	}
+	if cm.loop != nil {
+		cm.loop.Stop()
+		cm.loop = nil
+	}
+	cm.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+// Properties returns whether or not the motor supports certain optional properties.
+func (cm *controlledMotor) Properties(ctx context.Context, extra map[string]interface{}) (motor.Properties, error) {
+	return motor.Properties{
+		PositionReporting: true,
+	}, nil
+}
+
+// Position reports the position of the motor based on its encoder. If it's not supported, the returned
+// data is undefined. The unit returned is the number of revolutions which is intended to be fed
+// back into calls of GoFor.
+func (cm *controlledMotor) Position(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	ticks, _, err := cm.enc.Position(ctx, encoder.PositionTypeTicks, extra)
+	if err != nil {
+		return 0, err
+	}
+
+	// offsetTicks in Rotation can be changed by ResetPosition
+	cm.mu.RLock()
+	defer cm.mu.RUnlock()
+	return (ticks + cm.offsetInTicks) / cm.ticksPerRotation, nil
+}
+
+// ResetZeroPosition sets the current position (+/- offset) to be the new zero (home) position.
+func (cm *controlledMotor) ResetZeroPosition(ctx context.Context, offset float64, extra map[string]interface{}) error {
+	// Stop the motor if resetting position
+	if err := cm.Stop(ctx, extra); err != nil {
+		return err
+	}
+	if err := cm.enc.ResetPosition(ctx, extra); err != nil {
+		return err
+	}
+
+	cm.mu.Lock()
```

Comment:
other modifications to `cm.offsetInTicks` use RLock, not sure if this should be the same? I don't know the distinction

Reply:
I _think_ RLock means just for reading and in this case we are writing to `cm.offsetInTicks`, so I assume that's why it's different